### PR TITLE
[WIP] [3.5] Feature alarm assigning

### PR DESCRIPTION
**Pull request by imbeacon**
## Pull Request description

Contains back end for alarm assigning feature and tests.

After adding this feature - users will be allowed to assign alarm to themselves or to other users. Also current implementation has ability to search alarms using assigneeId as a filter.
Expected changes in documentation - adding information about how to assign alarm, get alarms for assigned user, check audit logs to get information when alarm assignment was changed.

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: application/src/main/data/upgrade/3.4.3/schema_update.sql [hunk 1]
```diff
@@ -14,6 +14,16 @@
 -- limitations under the License.
 --
 
+
+-- ALARM ASSIGN TO USER START
+
+ALTER TABLE alarm ADD COLUMN IF NOT EXISTS assign_ts BIGINT;
+ALTER TABLE alarm ADD COLUMN IF NOT EXISTS assignee_id UUID;
+
+ALTER TABLE entity_alarm ADD COLUMN IF NOT EXISTS assignee_id UUID;
```
Comment (ashvayka) [inline application/src/main/data/upgrade/3.4.3/schema_update.sql:23]: The `entity_alarm` relation stores the relationship between the alarm and the propagated entity. I don't we need `assignee_id` there.

File: application/src/main/java/org/thingsboard/server/controller/AlarmController.java [hunk 2]
```diff
@@ -179,6 +188,51 @@ public void clearAlarm(@ApiParam(value = ALARM_ID_PARAM_DESCRIPTION) @PathVariab
...
+        checkParameter(ALARM_ID, strAlarmId);
+        checkParameter(ASSIGNEE_ID, strAssigneeId);
+        AlarmId alarmId = new AlarmId(toUUID(strAlarmId));
+        UserId assigneeId = new UserId(UUID.fromString(strAssigneeId));
+        Alarm alarm = checkAlarmId(alarmId, Operation.WRITE);
+        User assigneeUser = userService.findUserById(getTenantId(), assigneeId);
+
+        SecurityUser assigneeSecurityUser = new SecurityUser(assigneeUser, false, null);
+        try {
+            checkEntityId(alarm.getOriginator(), Operation.WRITE, assigneeSecurityUser);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/AlarmController.java:213]: This check is redundant. We just need to check that the current user has access to the alarm and the assignee belongs to the same tenant.

File: application/src/main/java/org/thingsboard/server/controller/BaseController.java [hunk 3]
```diff
@@ -458,11 +458,15 @@ protected SecurityUser getCurrentUser() throws ThingsboardException {
     }
 
     Tenant checkTenantId(TenantId tenantId, Operation operation) throws ThingsboardException {
+        return checkTenantId(tenantId, operation, getCurrentUser());
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/BaseController.java:461]: Please revert all these changes and better recreate the PR without those changes in the commit history. This is completely not needed.

File: application/src/main/java/org/thingsboard/server/service/action/EntityActionService.java [hunk 4]
```diff
@@ -86,6 +86,12 @@ public void pushEntityActionToRuleEngine(EntityId entityId, HasName entity, Tena
             case ALARM_CLEAR:
                 msgType = DataConstants.ALARM_CLEAR;
                 break;
+            case ALARM_ASSIGN:
+                msgType = DataConstants.ALARM_CLEAR;
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/action/EntityActionService.java:90]: Why ALARM_CLEAR?

File: application/src/main/java/org/thingsboard/server/service/action/EntityActionService.java [hunk 5]
```diff
@@ -86,6 +86,12 @@ public void pushEntityActionToRuleEngine(EntityId entityId, HasName entity, Tena
...
+                break;
+            case ALARM_UNASSIGN:
+                msgType = DataConstants.ALARM_CLEAR;
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/action/EntityActionService.java:93]: Why ALARM_CLEAR?

File: application/src/main/java/org/thingsboard/server/service/entitiy/alarm/DefaultTbAlarmService.java [hunk 6]
```diff
@@ -93,6 +94,34 @@ public ListenableFuture<Void> clear(Alarm alarm, User user) {
         }, MoreExecutors.directExecutor());
     }
 
+    @Override
+    public ListenableFuture<Void> assign(Alarm alarm, User user, UserId assigneeId) {
+        long assignTs = System.currentTimeMillis();
+        ListenableFuture<Boolean> future = alarmSubscriptionService.assignAlarm(alarm.getTenantId(), alarm.getId(), assigneeId, assignTs);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/entitiy/alarm/DefaultTbAlarmService.java:100]: This should return saved alarm. Then we should use the saved alarm for notification. We can't duplicate: 
 
```
                alarm.setAssignTs(assignTs);
                alarm.setAssigneeId(assigneeId);
```

File: application/src/main/java/org/thingsboard/server/service/entitiy/alarm/TbAlarmService.java [hunk 7]
```diff
@@ -28,5 +29,9 @@ public interface TbAlarmService {
 
     ListenableFuture<Void> clear(Alarm alarm, User user);
 
+    ListenableFuture<Void> assign(Alarm alarm, User user, UserId assigneeId);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/entitiy/alarm/TbAlarmService.java:32]: No need for ListenableFuture. This operation goes to the PostgreSQL DB, and it is not async. Use similar pattern that we use in ```Alarm save(Alarm entity, User user)```

File: application/src/main/java/org/thingsboard/server/service/entitiy/alarm/DefaultTbAlarmService.java [hunk 8]
```diff
@@ -93,6 +94,34 @@ public ListenableFuture<Void> clear(Alarm alarm, User user) {
...
+        }, MoreExecutors.directExecutor());
+    }
+
+    @Override
+    public ListenableFuture<Void> unassign(Alarm alarm, User user) {
+        long assignTs = System.currentTimeMillis();
+        ListenableFuture<Boolean> future = alarmSubscriptionService.unassignAlarm(alarm.getTenantId(), alarm.getId(), assignTs);
+        return Futures.transform(future, result -> {
+            if (result != null && result) {
+                alarm.setAssignTs(assignTs);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/entitiy/alarm/DefaultTbAlarmService.java:117]: same anti-pattern here.

File: common/cluster-api/src/main/proto/queue.proto [hunk 9]
```diff
@@ -580,7 +580,7 @@ message TbAlarmUpdateProto {
   int64 entityIdLSB = 3;
   int64 tenantIdMSB = 4;
   int64 tenantIdLSB = 5;
-  string alarm = 6;
+  string alarmInfo = 6;
```
Comment (ashvayka) [inline common/cluster-api/src/main/proto/queue.proto:583]: Why change this if the AlarmInfo extends the Alarm?

File: application/src/main/java/org/thingsboard/server/controller/AlarmController.java [hunk 10]
```diff
@@ -179,6 +188,48 @@ public void clearAlarm(@ApiParam(value = ALARM_ID_PARAM_DESCRIPTION) @PathVariab
...
+                            @PathVariable(ALARM_ID) String strAlarmId,
+                            @ApiParam(value = ASSIGN_ID_PARAM_DESCRIPTION)
+                            @PathVariable(ASSIGNEE_ID) String strAssigneeId
+    ) throws Exception {
+        checkParameter(ALARM_ID, strAlarmId);
+        checkParameter(ASSIGNEE_ID, strAssigneeId);
+        AlarmId alarmId = new AlarmId(toUUID(strAlarmId));
+        Alarm alarm = checkAlarmId(alarmId, Operation.WRITE);
+        UserId assigneeId = new UserId(UUID.fromString(strAssigneeId));
+        User user = userService.findUserById(getCurrentUser().getTenantId(), assigneeId);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/AlarmController.java:209]: checkUserId(READ)

File: application/src/main/java/org/thingsboard/server/service/telemetry/sub/AlarmSubscriptionUpdate.java [hunk 11]
```diff
@@ -36,18 +37,18 @@ public class AlarmSubscriptionUpdate {
     @Getter
     private String errorMsg;
     @Getter
-    private Alarm alarm;
+    private AlarmInfo alarmInfo;
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/telemetry/sub/AlarmSubscriptionUpdate.java:40]: Please DO NOT rename the fields that are part of the public API.

**Comment by imbeacon** [comment]
Squashed - https://github.com/thingsboard/thingsboard/pull/8086

**Pull request closed without merging**